sbat.c

```diff
@@ -294,4 +294,96 @@ parse_sbat_var(list_t *entries)
 	return parse_sbat_var_data(entries, data, datasize+1);
 }
 
+EFI_STATUS
+set_sbat_uefi_variable(void)
+{
+	EFI_STATUS efi_status = EFI_SUCCESS;
+	UINT32 attributes = 0;
+
+	CHAR8 *sbat = NULL;
+	UINTN sbatsize = 0;
+
+	/* get size */
+	efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID, &attributes,
+			&sbatsize, sbat);
+	if (EFI_ERROR(efi_status))
+		dprint(L"SBAT read failed %r\n", efi_status);
+
+	sbat = AllocateZeroPool(sbatsize);
+	if (!sbat)
+		return EFI_OUT_OF_RESOURCES;
+
+	/* read and get attributes */
+	efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID, &attributes,
+			&sbatsize, sbat);
+	if (EFI_ERROR(efi_status))
+		dprint(L"SBAT read failed %r\n", efi_status);
+
+	/*
+	 * If attributes and the initial part of the version match, then
+	 * there is no need to re-write it. If it is not set or the read
+	 * failed, then attributes won't match.
+	 *
+	 * Finally, if the SBAT_VAR built into this shim is newer, update
+	 * it. Either the sbat version or date can be newer, and since
+	 * the version is more significant than the date, strcmp() applies.
+	 */
+	if (((attributes != (EFI_VARIABLE_NON_VOLATILE |
+		EFI_VARIABLE_BOOTSERVICE_ACCESS))
+		|| (attributes != (EFI_VARIABLE_NON_VOLATILE |
+				EFI_VARIABLE_BOOTSERVICE_ACCESS |
+				EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS)))
+		|| (strncmp(sbat, SBAT_VAR_SIG, strlen(SBAT_VAR_SIG)) != 0)
+		|| (strncmp(SBAT_VAR, sbat, strlen(SBAT_VAR)) > 0)
+		) {
+
+		/* delete previous variable */
+		efi_status = gRT->SetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				attributes, 0, "");
+		if (EFI_ERROR(efi_status))
+			dprint(L"SBAT variable delete failed %r\n", efi_status);
+
+		/* verify that it's gone */
+		efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				0, &sbatsize, sbat);
+		if ((efi_status != EFI_SUCCESS) || (sbatsize != 0))
+			dprint(L"SBAT variable clearing failed %r\n", efi_status);
+
+		/* set variable */
+		efi_status = gRT->SetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				EFI_VARIABLE_NON_VOLATILE |
+				EFI_VARIABLE_BOOTSERVICE_ACCESS,
+				sizeof (SBAT_VAR), SBAT_VAR);
+		if (EFI_ERROR(efi_status))
+			dprint(L"SBAT variable writing failed %r\n", efi_status);
+
+		/* verify that the expected data is there */
+		efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				0, &sbatsize, NULL);
+		if (EFI_ERROR(efi_status))
+			dprint(L"SBAT read failed %r\n", efi_status);
+
+		/* re-allocate in case of HV intervention  */
```

Comment:
I have no idea what this comment means.  We need to reallocate anyway in order to fit our new variable, which is most likely larger.  Maybe nix the comment in favor of saying so?

Reply:
Done and yes, the primary point of that action is clear.